src/ucp/core/ucp_worker.c

```diff
@@ -1169,6 +1169,76 @@ void ucp_worker_iface_cleanup(ucp_worker_iface_t *wiface)
     ucp_worker_uct_iface_close(wiface);
 }
 
+static ucs_status_t ucp_worker_add_resource_cms(ucp_worker_h worker)
+{
+    ucp_context_h   context  = worker->context;
+    ucs_status_t    status;
+    ucp_rsc_index_t i, j;
+
+    if (!ucp_worker_sockaddr_is_cm_proto(worker)) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    UCS_ASYNC_BLOCK(&worker->async);
+    worker->num_cms = 0;
+    for (i = 0; i < context->num_cmpts; ++i) {
```

Comment:
count this when context is created and keep num_cms on context
instead of ucp_worker_sockaddr_is_cm_proto() function, do (worker->num_cms > 0)

Reply:
did you mean `worker->context->num_cms`? Anyway I would prefer to have a func for this..

Reply:
`worker->context->num_cms` won't work because this is not an components, this is CM objects which will be opened in scope of this worker

Reply:
i see that worker->num_cms counts components with UCT_COMPONENT_FLAG_CM flag. if any cm_open function fails, it's fatal anyway

Reply:
why is it fatal? it cleans up cms and returns error, this logic might be changed if we add several components for the same fabric. I would prefer to have this on the same level.

Reply:
can replace it with `ucs_popcount(context->cm_tls_bitmap)` according to your comment below, wdyt?